python/demo/gsoc2022/demo_scattering_boundary_conditions.py

```diff
@@ -0,0 +1,451 @@
+
+MPI.COMM_WORLD.barrier()
+
+# Definition of finite element for the electric field
+curl_el = ufl.FiniteElement("N1curl", mesh.ufl_cell(), 3)
+V = fem.FunctionSpace(mesh, curl_el)
+
+# Wavevector of the background field
+k0 = 2 * np.pi / wl0
+
+# Angle of incidence of the background field
+deg = np.pi / 180
+theta = 45 * deg
+
+# Plane wave function
+f = background_electric_field(theta, n_bkg, k0)
+Eb = fem.Function(V)
+Eb.interpolate(f.eval)
+
+# Function r = radial distance from the (0, 0) point
+lagr_el = ufl.FiniteElement("CG", mesh.ufl_cell(), 2)
+lagr_space = fem.FunctionSpace(mesh, lagr_el)
+r = fem.Function(lagr_space)
+r.interpolate(radial_distance)
+
+# Definition of Trial and Test functions
+Es = ufl.TrialFunction(V)
+Vs = ufl.TestFunction(V)
+
+# Definition of 3d fields for cross and curl operations
+Es_3d = as_vector((Es[0], Es[1], 0))
+Vs_3d = as_vector((Vs[0], Vs[1], 0))
+
+# Measures for subdomains
+dx = ufl.Measure("dx", mesh, subdomain_data=cell_tags)
+ds = ufl.Measure("ds", mesh, subdomain_data=facet_tags)
+dAu = dx(au_tag)
+dBkg = dx(bkg_tag)
+dDom = dAu + dBkg
+dsbc = ds(boundary_tag)
+
+# Normal to the boundary
+n = FacetNormal(mesh)
+n_3d = as_vector((n[0], n[1], 0))
+
+# Definition of relative permittivity for Au @400nm
+reps_au = -1.0782
+ieps_au = 5.8089
+eps_au = reps_au + ieps_au * 1j
+
+# Definition of the relative permittivity over the whole domain
+D = fem.FunctionSpace(mesh, ("DG", 0))
+eps = fem.Function(D)
+au_cells = cell_tags.find(au_tag)
+bkg_cells = cell_tags.find(bkg_tag)
+eps.x.array[au_cells] = np.full_like(
+    au_cells, reps_au + ieps_au * 1j, dtype=np.complex128)
+eps.x.array[bkg_cells] = np.full_like(bkg_cells, eps_bkg, dtype=np.complex128)
+
+# Weak form
+F = - inner(curl(Es), curl(Vs)) * dDom \
+    + eps * k0 ** 2 * inner(Es, Vs) * dDom \
+    + k0 ** 2 * (eps - eps_bkg) * inner(Eb, Vs) * dDom \
+    + (1j * k0 * n_bkg + 1 / (2 * r)) \
+    * inner(cross(Es_3d, n_3d), cross(Vs_3d, n_3d)) * dsbc
+
+# Splitting in left-hand side and right-hand side
+a, L = lhs(F), rhs(F)
+
+problem = fem.petsc.LinearProblem(a, L, bcs=[], petsc_options={
+                                  "ksp_type": "preonly", "pc_type": "lu"})
+Eh = problem.solve()
+
+# Total electric field E = Es + Eb
+E = fem.Function(V)
+E.x.array[:] = Eb.x.array[:] + Eh.x.array[:]
+
+# ||E||
+norm_func = sqrt(inner(Eh, Eh))
+V_normEh = fem.FunctionSpace(mesh, lagr_el)
+norm_expr = fem.Expression(norm_func, V_normEh.element.interpolation_points)
+normEh = fem.Function(V_normEh)
+normEh.interpolate(norm_expr)
+
+# Save the fields as xdmf files
+with io.XDMFFile(MPI.COMM_WORLD, "data/Es.xdmf", "w") as xdmf:
+    xdmf.write_mesh(mesh)
+    xdmf.write_function(Eh)
+
+with io.XDMFFile(MPI.COMM_WORLD, "data/E.xdmf", "w") as xdmf:
+    xdmf.write_mesh(mesh)
+    xdmf.write_function(E)
+
+with io.XDMFFile(MPI.COMM_WORLD, "data/normEs.xdmf", "w") as xdmf:
+    xdmf.write_mesh(mesh)
+    xdmf.write_function(normEh)
```

Comment:
As we are using higher order Nedelec functions, we should use `VTXWriter` and first interpolate the fields into a suitable (order 3) vector DG space.